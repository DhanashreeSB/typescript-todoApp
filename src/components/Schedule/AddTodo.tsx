import React, { useState } from 'react'
import useStores from '../../hooks/useStores';

const AddTodo = () => {

    const [id, setId] = useState(0);
    const [content, setContent] = useState('');
    const { todoState } = useStores();

    const handleChange = (e: any) => {
        setContent(e.target.value);
    }

    const handleSubmit = (e: any) => {
        e.preventDefault();
        todoState.addTodo({ id, content })
        //To make the text field empty after submitting one task
        setId(0);
        setContent('');
    }

    return (
        <div>
            <form onSubmit={(e) => handleSubmit(e)}>
                <label className="left black-text">Add new Todo:</label>
                <input type="text" onChange={handleChange} value={content} data-testid="add-todo"/>
                <button type="submit" className="center green white-text" id="add" >Add Task</button>
            </form>
        </div>
    )
}

export default AddTodo