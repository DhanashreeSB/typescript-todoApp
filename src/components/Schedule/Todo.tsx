import React from 'react'
import useStores from '../../hooks/useStores';
import { workTodoI } from '../../models/todos.interface';

type Props = {
    todos: Array<workTodoI>;
}

const Todo = (props: Props) => {
    const { todoState } = useStores();

    //display each element using map:
    const todoList = props.todos.length ? (
        props.todos.map((todo, index) => {
            return (
                <div className="collection-item" key={todo.id}>
                    <span>{todo.id}. {todo.content}</span>
                    <button className="right green white-text" data-testid={`todo-${index}`} onClick={() => todoState.deleteTodos(todo.id)}>Remove</button>
                </div>
            )
        })

    ) : (<p className="center">You have no todo's left</p>)
    return (
        <div className="todos collection">
            {todoList}
        </div>
    )
}

export default Todo;