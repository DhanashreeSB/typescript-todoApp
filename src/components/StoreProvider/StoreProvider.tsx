import React from 'react';
import { createRootStore } from '../../stores/RootStore';

export const storeContext = React.createContext(null);

export const StoreProvider: React.FC = ({children}) => {
    const store = createRootStore();

    return (
        /* @ts-ignore */
        <storeContext.Provider value={store}>{children}</storeContext.Provider>
    )
}