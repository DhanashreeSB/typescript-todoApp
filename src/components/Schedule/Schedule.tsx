import React, { useState } from 'react';
import Todo from './Todo';
import AddTodo from './AddTodo';
import { observer } from 'mobx-react';
import useStores from '../../hooks/useStores';
import { Redirect } from 'react-router-dom';

const Schedule = observer(() => {

    const { todoState } = useStores();
    const [back,setBack] = useState(false);

    const handleBack = () => {
        setBack(true);
    }

    if(back) {
        return(<Redirect to="/"/>)
    }

    return (
        <div className="todo-app container" id="todo-list">
            <h2 className="center blue-text">To do</h2>
            <Todo todos={todoState.todos} />
            <AddTodo />
            <p onClick={handleBack} style={{color:"blue", cursor: "pointer"}}>Go to home</p>
        </div>
    )
})

export default Schedule