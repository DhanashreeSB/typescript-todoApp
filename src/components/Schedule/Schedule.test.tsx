import React from 'react';
import { render, screen } from '@testing-library/react';
// import * as todoConstants from '../../Constants/todoConstants';
import Schedule from './Schedule';
import { StoreProvider } from '../StoreProvider/StoreProvider';
// import useStores from '../../hooks/useStores';
import userEvent from '@testing-library/user-event'

jest.mock('react-router-dom');
// jest.mock("./Todo")

describe("Test Home component", () => {
    test("snapshot testing", () => {
        let snapshot = render(<StoreProvider><Schedule /></StoreProvider>);
        expect(snapshot).toMatchSnapshot();
    })
    describe("When user clicks on Go to home", () => {
        beforeEach(async () => {
            const mockRouterDom = require('react-router-dom');
            mockRouterDom.Redirect = jest.fn((props) => { return <div>{props.to}</div> });
            render(<StoreProvider><Schedule /></StoreProvider>);
            await (await screen.findByText("Go to home")).click();
        })
        it("Should redirect to /", () => {
            expect(screen.getByText("/")).toBeInTheDocument();
        })
    })
    describe("Check for initial todos",()=>{
        beforeEach(()=>{
            render(<StoreProvider><Schedule /></StoreProvider>);
        })
        it("should show initial todos on screen",()=>{
            expect(screen.getByText(/Meeting at 3PM/i)).toBeInTheDocument();
            expect(screen.getByText(/Presentation at 1PM/i)).toBeInTheDocument();
        })
        describe("Given user clicks on remove button",()=>{
            beforeEach(()=>{
                let node = screen.getByTestId("todo-0");
                node.click();
            })
            it("should remove one todo from screen",()=>{
                let todo = screen.queryByText(/Presentation at 1PM/i)
                expect(todo).toBeNull();
            })
        })
        describe("Given user enters new work item and clicks on add button",()=>{
            beforeEach(()=>{
                let node = screen.getByTestId("add-todo");
                userEvent.type(node, "New work todo");
                let addTodo = screen.getByText("Add Task");
                addTodo.click();
            })
            it("should show newly added todo on screen",()=>{
                expect(screen.getByText(/New work todo/i)).toBeInTheDocument();
            })
        })
    })
})