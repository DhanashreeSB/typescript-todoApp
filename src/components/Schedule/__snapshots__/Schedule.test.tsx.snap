// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Home component snapshot testing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="todo-app container"
        id="todo-list"
      >
        <h2
          class="center blue-text"
        >
          To do
        </h2>
        <div
          class="todos collection"
        >
          <div
            class="collection-item"
          >
            <span>
              1
              . 
              Presentation at 1PM
            </span>
            <button
              class="right green white-text"
              data-testid="todo-0"
            >
              Remove
            </button>
          </div>
          <div
            class="collection-item"
          >
            <span>
              2
              . 
              Meeting at 3PM
            </span>
            <button
              class="right green white-text"
              data-testid="todo-1"
            >
              Remove
            </button>
          </div>
        </div>
        <div>
          <form>
            <label
              class="left black-text"
            >
              Add new Todo:
            </label>
            <input
              data-testid="add-todo"
              type="text"
              value=""
            />
            <button
              class="center green white-text"
              id="add"
              type="submit"
            >
              Add Task
            </button>
          </form>
        </div>
        <p
          style="color: blue; cursor: pointer;"
        >
          Go to home
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="todo-app container"
      id="todo-list"
    >
      <h2
        class="center blue-text"
      >
        To do
      </h2>
      <div
        class="todos collection"
      >
        <div
          class="collection-item"
        >
          <span>
            1
            . 
            Presentation at 1PM
          </span>
          <button
            class="right green white-text"
            data-testid="todo-0"
          >
            Remove
          </button>
        </div>
        <div
          class="collection-item"
        >
          <span>
            2
            . 
            Meeting at 3PM
          </span>
          <button
            class="right green white-text"
            data-testid="todo-1"
          >
            Remove
          </button>
        </div>
      </div>
      <div>
        <form>
          <label
            class="left black-text"
          >
            Add new Todo:
          </label>
          <input
            data-testid="add-todo"
            type="text"
            value=""
          />
          <button
            class="center green white-text"
            id="add"
            type="submit"
          >
            Add Task
          </button>
        </form>
      </div>
      <p
        style="color: blue; cursor: pointer;"
      >
        Go to home
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
